contracts/SponsorFaucet.sol

```diff
@@ -64,38 +109,180 @@ contract SponsorFaucet is Ownable, Pausable, ReentrancyGuard {
     function isAppliable(address dapp) public returns (bool) {
         if (_isAppliableForGas(dapp) || _isAppliableForCollateral(dapp))
             return true;
+        require(dapp.isContract(), "ERROR_ADDRESS_IS_NOT_CONTRACT");
         _validateApplyForGas(dapp);
         _validateApplyForCollateral(dapp);
     }
 
+    //accept sponsor's cfx
+    function() external payable {}
+
+    //withdraw to specific address by amount
+    function withdraw(address payable sponsor, uint256 amount)
+        public
+        onlyOwner
+        nonReentrant
+        whenPaused
+    {
+        require(address(this).balance >= amount, "amount too high");
+        (bool success, ) = sponsor.call.value(amount)("");
+        require(success, "withdraw failed");
+    }
+
+    //set bounds for sponsorship
+    function setBounds(
+        uint256 gasTotalLimit,
+        uint256 collateralTotalLimit,
+        uint256 gasBound,
+        uint256 collateralBound,
+        uint256 upperBound
+    ) public {
+        _onlyWhitelistAdmin();
+        //rule by internal contract
+        require(upperBound.mul(1000) <= gasBound, "upperBound too high");
+        gas_total_limit = gasTotalLimit;
+        collateral_total_limit = collateralTotalLimit;
+        gas_bound = gasBound;
+        collateral_bound = collateralBound;
+        upper_bound = upperBound;
+    }
+
+    //set bounds for general contracts
+    function setGeneralBounds(
+        uint256 gasTotalLimit,
+        uint256 collateralTotalLimit,
+        uint256 gasBound,
+        uint256 collateralBound,
+        uint256 upperBound
+    ) public {
+        _onlyWhitelistAdmin();
+        //rule by internal contract
+        require(upperBound.mul(1000) <= gasBound, "upperBound too high");
+        general_gas_total_limit = gasTotalLimit;
+        general_collateral_total_limit = collateralTotalLimit;
+        general_gas_bound = gasBound;
+        general_collateral_bound = collateralBound;
+        general_upper_bound = upperBound;
+    }
+
+    //set bounds for special contracts
+    function setSpecialBounds(
+        address addr,
+        uint256 gasTotalLimit,
+        uint256 collateralTotalLimit,
+        uint256 gasBound,
+        uint256 collateralBound,
+        uint256 upperBound
+    ) public {
+        _onlyWhitelistAdmin();
+        //rule by internal contract
+        require(upperBound.mul(1000) <= gasBound, "upperBound too high");
+        special_dapps[addr].gas_total_limit = gasTotalLimit;
+        special_dapps[addr].collateral_total_limit = collateralTotalLimit;
+        special_dapps[addr].gas_bound = gasBound;
+        special_dapps[addr].collateral_bound = collateralBound;
+        special_dapps[addr].upper_bound = upperBound;
+    }
+
+
+    /* ===== Public utility functions ===== */
+    function addGeneralContracts(address[] memory addAddr) public {
```

Comment:
Why not let general contracts become a case of special contract?

Reply:
Different mapping **large_contract** and **custom_contracts** are needed to be updated in **add functions**.